diff --git a/pysewer/config/settings.py b/pysewer/config/settings.py
index 4af89c9..c3d2bf8 100644
--- a/pysewer/config/settings.py
+++ b/pysewer/config/settings.py
@@ -36,7 +36,8 @@ class Preprocessing:
 
 @dataclass
 class Optimization:
-    inhabitants_dwelling: int
+    inhabitants_dwelling_attribute_name: str
+    default_inhabitants_dwelling: int
     daily_wastewater_person: float
     peak_factor: float
     min_slope: float
diff --git a/pysewer/config/settings.yaml b/pysewer/config/settings.yaml
index ca6ee25..4fddd23 100644
--- a/pysewer/config/settings.yaml
+++ b/pysewer/config/settings.yaml
@@ -20,7 +20,8 @@ preprocessing:
   value_get_buildings: "building" # value for buildings
 
 optimization:
-  inhabitants_dwelling: 3
+  default_inhabitants_dwelling: 3
+  inhabitants_dwelling_attribute_name: "inhabitants_nb"
   daily_wastewater_person: 0.2 # m³/person/day
   peak_factor: 2.3 # peak factor for wastewater
   min_slope: -0.01 # min slope required for gravitational flow
diff --git a/pysewer/export.py b/pysewer/export.py
index c461f98..7f46de2 100644
--- a/pysewer/export.py
+++ b/pysewer/export.py
@@ -11,9 +11,10 @@ The option saving the data as a parquet file is added.
 """
 
 import json
-from typing import List, Tuple
+from typing import List, Tuple, Optional
 
 import fiona
+from fiona.collection import CRS
 import geopandas as gpd
 
 from .config.settings import load_config
@@ -27,6 +28,8 @@ def map_dtype_to_fiona(dtype):
         return "int"
     elif "float" in dtype:
         return "float"
+    elif "bool" in dtype:
+        return "bool"
     else:
         return "str"
 
@@ -39,6 +42,9 @@ def is_list_of_tuples(column):
         for item in column
     )
 
+def is_list(column):
+    """Check if a pandas Series contains lists."""
+    return all(isinstance(item, list) for item in column)
 
 def tuple_list_to_json(tuple_list):
     """Serialize a list of tuples to a JSON string."""
@@ -71,7 +77,7 @@ def generate_schema(gdf: gpd.GeoDataFrame):
     return schema
 
 
-def write_gdf_to_gpkg(gdf: gpd.GeoDataFrame, filepath: str):
+def write_gdf_to_gpkg(gdf: gpd.GeoDataFrame, filepath: str, layer: Optional[str]=None, crs: Optional[CRS]=None):
     """
     Write a GeoDataFrame to a GeoPackage (GPKG) file using Fiona, converting lists of tuples to JSON strings.
 
@@ -81,6 +87,10 @@ def write_gdf_to_gpkg(gdf: gpd.GeoDataFrame, filepath: str):
         The GeoDataFrame to be written to the GPKG file.
     filepath : str
         The file path to the GPKG file.
+    layer: str (optional)
+        The layer name inside the GPKG file. If no layer name is specified the file name is used.
+    crs: CRS (optional)
+        The layer CRS
 
     Returns
     -------
@@ -95,12 +105,14 @@ def write_gdf_to_gpkg(gdf: gpd.GeoDataFrame, filepath: str):
         if gdf[col].dtype == "object":
             if is_list_of_tuples(gdf[col]):
                 gdf[col] = gdf[col].apply(tuple_list_to_json)
+            if is_list(gdf[col]):
+                gdf[col] = gdf[col].apply(str)
 
     # Define the schema based on the GeoDataFrame
     schema = generate_schema(gdf)
 
     # Open a new GPKG file in write mode
-    with fiona.open(filepath, mode="w", driver="GPKG", schema=schema) as dst:
+    with fiona.open(filepath, mode="w", driver="GPKG", schema=schema, layer=layer, crs=crs) as dst:
         for _, row in gdf.iterrows():
             feature = {
                 "geometry": row["geometry"].__geo_interface__,
diff --git a/pysewer/optimization.py b/pysewer/optimization.py
index 7c3887c..58c0f45 100644
--- a/pysewer/optimization.py
+++ b/pysewer/optimization.py
@@ -266,10 +266,16 @@ def calculate_hydraulic_parameters(
                     edge: {
                         "trench_depth_profile": td_profile,
                         "mean_td": np.mean(
-                            [topo[1] - td[1] for td, topo in zip(td_profile, profile)]
+                            [topo[1] - td[1] for td, topo in zip(td_profile, [profile[0] + profile[-1]])]
                         ),
                     }
                 }
+                node_attrs = {
+                    upstream: {
+                        "total_static_head": round(td_profile[1][1] - td_profile[0][1] + max(0, max_inflow_tds - DEFAULT_CONFIG.optimization.tmin),  2)
+                    }
+                }
+                nx.set_node_attributes(G, node_attrs)
 
             else:
                 # check if edge needs a lifting station based on max inflow td
@@ -282,13 +288,16 @@ def calculate_hydraulic_parameters(
                     _, out_trench_depth, td_profile = needs_pump(
                         profile=profile, inflow_trench_depth=min_trench_depth
                     )
-                # append downstream inflow td value
+                # append downstream inflow td value and upstream total static head
                 node_attrs = {
                     downstream: {
                         "inflow_trench_depths": G.nodes()[downstream][
                             "inflow_trench_depths"
                         ]
                         + [out_trench_depth]
+                    },
+                    upstream: {
+                        "total_static_head": round(max_inflow_tds - DEFAULT_CONFIG.optimization.tmin, 2)
                     }
                 }
                 nx.set_node_attributes(G, node_attrs)
@@ -346,7 +355,9 @@ def calculate_hydraulic_parameters(
 
 def estimate_peakflow(
     G: nx.Graph,
-    inhabitants_dwelling: int = DEFAULT_CONFIG.optimization.inhabitants_dwelling,
+    # inhabitants_dwelling: int = DEFAULT_CONFIG.optimization.inhabitants_dwelling,
+    inhabitants_dwelling_attribute_name: str = DEFAULT_CONFIG.optimization.inhabitants_dwelling_attribute_name,
+    default_inhabitants_dwelling: int = DEFAULT_CONFIG.optimization.default_inhabitants_dwelling,
     daily_wastewater_person: float = DEFAULT_CONFIG.optimization.daily_wastewater_person,
     peak_factor: float = DEFAULT_CONFIG.optimization.peak_factor,
 ):
@@ -357,8 +368,10 @@ def estimate_peakflow(
     ----------
     G : networkx.Graph
         The graph to estimate peakflow for.
-    inhabitants_dwelling : int
-        The number of inhabitants per dwelling.
+    inhabitants_dwelling_attribute_name : str
+        The attribute name with the number of inhabitants per dwelling.
+    default_inhabitants_dwelling : int
+        The default number of inhabitants per dwelling to use if inhabitants_dwelling_attribute_name is empty
     daily_wastewater_person : float
         The daily wastewater generated per person in m³.
     peak_factor : float, optional
@@ -371,10 +384,16 @@ def estimate_peakflow(
     """
     for n in G.nodes():
         upstream_buildings = get_upstream_nodes(G, n, "node_type", "building")
-        upstream_daily = (
-            len(upstream_buildings) * inhabitants_dwelling * daily_wastewater_person
-        )
-        upstream_pe = len(upstream_buildings) * inhabitants_dwelling
+
+        if inhabitants_dwelling_attribute_name != "":
+            upstream_pe = 0
+            for upstream_building in upstream_buildings:
+                inhabitant_number = float(G.nodes[upstream_building][inhabitants_dwelling_attribute_name])
+                upstream_pe += inhabitant_number
+        else:
+            upstream_pe = len(upstream_buildings) * default_inhabitants_dwelling
+        upstream_daily = upstream_pe * daily_wastewater_person
+        
         peak_flow = ((upstream_daily / 24) * peak_factor) / 3600
         atr = {
             n: {
@@ -518,9 +537,9 @@ def needs_pump(
                 [
                     (
                         0,
-                        tmin,
+                        y[0] - tmin,
                     ),
-                    (x[-1], tmin),
+                    (x[-1], y[-1] - tmin),
                 ],
             )
         ## case 2: min slope point is within Trench depth range: set trench depth to calculated height
diff --git a/pysewer/routing.py b/pysewer/routing.py
index 4bd3d18..b9358e4 100644
--- a/pysewer/routing.py
+++ b/pysewer/routing.py
@@ -145,17 +145,25 @@ def rsph_tree(
             break  # exit the loop if no viable terminal is found
 
         # Add edges of path to sewer graph while keeping edge attributes
+        # and propagate sink node in sewer graph attributes (edges and nodes)
         nx.add_path(sewer_graph, path)
 
         edgesinpath = zip(path[:-1], path[1:])
         for edge in edgesinpath:
-            nx.set_edge_attributes(
+            attrs = connection_graph.get_edge_data(edge[0], edge[1])[0]
+            path_dest_node = path[-1]
+            try:
+                sink_node = str(sewer_graph.nodes(data=True)[path_dest_node]["sink_coords"])
+            except KeyError:
+                sink_node = path_dest_node
+
+            attrs["sink_coords"] = str(sink_node)
+            nx.set_node_attributes(
                 sewer_graph,
-                {
-                    (edge[0], edge[1]): connection_graph.get_edge_data(
-                        edge[0], edge[1]
-                    )[0]
-                },
+                {edge_node: {"sink_coords": str(sink_node)} for edge_node in edge}
+            )
+            nx.set_edge_attributes(
+                sewer_graph, {(edge[0], edge[1]): attrs},
             )
         subgraph_nodes.extend(path[1:-1])
         subgraph_nodes = list(set(subgraph_nodes))  # Remove duplicates
